ethereum/statetransition/src/main/java/tech/pegasys/artemis/statetransition/StateTreeManager.java

```diff
@@ -74,47 +83,79 @@ public void onValidatorRegistered(ValidatorRegistrationEvent event) {
+            stateTransition.initiate(newState, null);
+            newStateRoot = HashTreeUtil.hash_tree_root(newState.toBytes());
+            LOG.info("  state slot: " + newState.getSlot());
+            LOG.info("  state root: " + newStateRoot.toHexString());
+            this.store.addState(newStateRoot, newState);
+            newState = BeaconState.deepCopy(newState);
+          }
+
           stateTransition.initiate(newState, block.get());
-          Bytes32 stateRoot = HashTreeUtil.hash_tree_root(newState.toBytes());
-          LOG.info("New state root: " + stateRoot.toHexString());
+          newStateRoot = HashTreeUtil.hash_tree_root(newState.toBytes());
+          LOG.info("  new state root: " + newStateRoot.toHexString());
+          LOG.info("  block state root: " + blockStateRoot.toHexString());
+
           // state root verification
           checkArgument(
-              blockStateRoot.equals(stateRoot),
+              blockStateRoot.equals(newStateRoot),
               "The block's state root %s does not match the calculated state root %s!",
               blockStateRoot.toHexString(),
-              stateRoot.toHexString());
+              newStateRoot.toHexString());
           // TODO: storing block and state together as a tuple would be more convenient
-          this.store.addState(stateRoot, newState);
+          this.store.addState(newStateRoot, newState);
           this.store.addProcessedBlock(blockStateRoot, block.get());
           this.store.addProcessedBlock(blockRoot, block.get());
         } else {
           LOG.info(
-              "Receieved a block without a parent.  Checking to see if state root matches this clients initial state root");
+              "Receieved genesis block.  Checking to see if state root matches this clients initial state root");
           Optional<BeaconState> state = this.store.getState(block.get().getState_root());
           if (state.isPresent()) {
             LOG.info("State roots match! Saving genesis block in storage");
             this.store.addProcessedBlock(blockStateRoot, block.get());
             this.store.addProcessedBlock(blockRoot, block.get());
+            newState = BeaconState.deepCopy(state.get());
           }
         }
-
-      } catch (NoSuchElementException | IllegalArgumentException | StateTransitionException e) {
-        LOG.warn(e);
+      } else {
+        // TODO: If there is no block for the current slot, should i just get the canonical state?
+        newState = BeaconState.deepCopy(this.canonical_state);
+        stateTransition.initiate(newState, null);
       }
-    } else {
-      // TODO: If there is no block for the current slot, should i just get the canonical state?
-      // stateTransition.initiate(state, null);
+      this.canonical_state = newState;
+      Bytes32 stateRoot = HashTreeUtil.hash_tree_root(newState.toBytes());
+      this.store.addState(stateRoot, newState);
```

Comment:
Might this, i.e. the addState(), be a duplicate because you already save the state for the block.isPresent case?

Reply:
Good question.... I'll double check before merging. 🧐